src/stack_check.h

```diff
@@ -57,41 +66,105 @@ static int stack_check_fn(unsigned long *entries, unsigned int nr_entries, bool
 		func_size = mod_func_size;
 	}
 
-	for (i = 0; i < nr_entries; i++) {
-		int idx;
-		unsigned long address = entries[i];
+	idx = bsearch(func_addr, 0, NR_INTERFACE_FN - 1, address);
+	if (idx == -1)
+		return 0;
+	if (address < func_addr[idx] + func_size[idx])
+		return -EAGAIN;
+	return 0;
+}
 
-		idx = bsearch(func_addr, 0, NR_INTERFACE_FN - 1, address);
-		if (idx == -1)
-			continue;
-		if (address < func_addr[idx] + func_size[idx])
-			return -EAGAIN;
+#ifdef CONFIG_ARCH_STACKWALK
+struct stacktrace_cookie {
+	unsigned long	*store;
+	unsigned int	size;
+	unsigned int	skip;
+	unsigned int	len;
+};
+
+static bool stack_trace_consume_entry(void *cookie, unsigned long addr)
+{
+	struct stacktrace_cookie *c = cookie;
+
+	if (c->len >= c->size)
+		return false;
+
+	if (c->skip > 0) {
+		c->skip--;
+		return true;
 	}
+	c->store[c->len++] = addr;
+	return c->len < c->size;
+}
 
-	return 0;
+static unsigned int get_stack_trace(struct task_struct *tsk,
+		unsigned long *store, unsigned int size)
+{
+	struct stacktrace_cookie c = {
+		.store  = store,
+		.size   = size,
+		.skip   = 0
+	};
+
+	if (!try_get_task_stack(tsk))
+		return 0;
+
+	arch_stack_walk(stack_trace_consume_entry, &c, tsk, NULL);
+	put_task_stack(tsk);
+	return c.len;
 }
+#else
+extern void __save_stack_trace(struct stack_trace *, struct task_struct *,
+		struct pt_regs *, bool);
 
-/* This is basically copied from klp_check_stack */
-static int stack_check_task(struct task_struct *task, bool install)
+static unsigned int get_stack_trace(struct task_struct *tsk,
+                       unsigned long *store, unsigned int size)
 {
-	unsigned long entries[MAX_STACK_ENTRIES];
-	unsigned int nr_entries = 0;
+        struct stack_trace trace;
 
-#ifndef CONFIG_ARCH_STACKWALK
-	struct stack_trace trace;
+        trace.skip = 0;
+        trace.nr_entries = 0;
+        trace.max_entries = MAX_STACK_ENTRIES;
+        trace.entries = store;
 
-	trace.skip = 0;
-	trace.nr_entries = 0;
-	trace.max_entries = MAX_STACK_ENTRIES;
-	trace.entries = entries;
+	if (!try_get_task_stack(tsk))
+		return 0;
 
-	save_stack_trace_tsk(task, &trace);
-	nr_entries = trace.nr_entries;
-#else /* CONFIG_ARCH_STACKWALK */
-	nr_entries = stack_trace_save_tsk(task, entries, MAX_STACK_ENTRIES, 0);
-#endif /* CONFIG_ARCH_STACKWALK */
+	__save_stack_trace(&trace, tsk, NULL, false);
+	put_task_stack(tsk);
+	return trace.nr_entries;
+}
+#endif
+
+static bool is___schedule(unsigned long addr, bool install)
+{
+	if (!install && addr >= (unsigned long)__mod___schedule &&
+	    addr < (unsigned long)__mod___schedule + mod_schedule_size)
+		return true;
+	if (install && addr >= (unsigned long)__orig___schedule &&
+	    addr < (unsigned long)__orig___schedule + orig_schedule_size)
+		return true;
+	return false;
+}
+
```

Comment:
Maybe there have another way to skip __schedule. For example, the variable of address of __schedule can be set 0 or ~0, then after address-sorting, the variable is placed at far left or far right, and the bsearch can help to skip this function.

Reply:
The size can be set zero, too.

Reply:
Setting vm_func_size[NR___schedule] and mod_func_size[NR___schedule] to 0 is more like a general and efficient way to skip checking some functions. Bravo.